### PR TITLE
Add comprehensive React support with provider and hook

### LINKED ISSUE #5: Create react support
I was thinking to create react support for CozyEvent, 

Maybe, we can create a component as `<eventEmitter> ... </eventEmitter>`  and some react hooks like useEventListener, ..emitter.  something practical for regarding declarative approach and general use-case in react.

### PR DESCRIPTION
**Pull request by jeangq24**
### **Pull Request Summary**
This Pull Request introduces **React support** for CozyEvent, enabling seamless integration of event-driven communication in React applications. The implementation focuses on providing a developer-friendly API, robust lifecycle management, and clear documentation to simplify the use of events in React.

---

#### **1. React Integration**
- **`CozyEventProvider`:**
  - A React Context provider that shares a `CozyEvent` instance with all child components.
  - Supports both global and custom `CozyEvent` instances.
  - Ensures proper validation of the provided instance.

- **`useCozyEvent`:**
  - A custom React hook that allows components to subscribe to specific events.
  - Automatically manages event subscriptions and cleanups using React's lifecycle (`useEffect`).
  - Supports namespaces for better organization of events in complex applications.

#### **2. Documentation**
- Updated the README.md file with:
  - A detailed explanation of the React integration.
  - Practical examples, including basic usage, namespaces, and lifecycle management.
  - Best practices for using CozyEvent in React applications.
- Added inline comments in the source code to improve clarity and maintainability.

#### **3. Tests**
- Comprehensive tests added in `ReactSupport.test.tsx` to ensure the reliability of the React integration:
  - Validation of parameters (`eventName`, `callback`, `namespace`).
  - Proper subscription and unsubscription of events.
  - Handling of global and custom `CozyEvent` instances.
  - Dynamic updates to `eventName` and `callback`.
  - Edge cases, such as invalid inputs and rapid re-renders.
- All tests pass successfully, ensuring the stability of the implementation.

#### **4. Core Enhancements**
- Ensured compatibility between the core `CozyEvent` functionality and the new React integration.
- No breaking changes introduced; the core library remains backward-compatible.

---

### **Why This Change?**
This change provides a robust and developer-friendly way to use CozyEvent in React applications. It simplifies event-driven communication, improves modularity, and ensures proper lifecycle management. By introducing React-specific tools, CozyEvent becomes more versatile and easier to adopt in modern frontend projects.

---

### **How to Test**
1. Run the test suite to verify the implementation:
   ```bash
   npm test
   ```
2. Follow the examples in the README.md to integrate CozyEvent into a React application and test its functionality.

---

### **Screenshots/Demos**
Below is a screenshot showing the successful execution of all tests:
Captura desde 2025-03-30 00-09-42

### CODE REVIEW

**Comment by jeangq24** [comment]
- Automatically register instances in CozyEventProvider.
- Add `registerCozyEventInstance` and `getCozyEventInstanceById` for centralized instance management.
- Update README.md with examples and best practices for multiple instances and the registry.
- Add comprehensive tests for multiple instances and registry functionality.

Captura desde 2025-03-30 04-32-33

**Review by RecursiveVoid — CHANGES_REQUESTED** [review]
I couldn't review it fully, but i added some suggestion and questions.  After those changes, ill be reviewing once again. Thanks for the changes! looks like a we gonna have a huge update! :D

File: src/react/CozyEventProvider.tsx [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+  children,
+  id = 'default', // Default ID for the provider
+}) => {
+  // Validate the instance
+  if (!(instance instanceof CozyEvent)) {
+    throw new Error('Invalid CozyEvent instance provided to CozyEventProvider');
+  }
+
+   // Register the instance with its ID
+   useMemo(() => {
```
[the remark this answers is not in the document]
Comment (jeangq24) [reply src/react/CozyEventProvider.tsx:31]: Good observation. UseEffect seems like the best option. It offers a built-in cleanup mechanism, aligns with React's lifecycle management, and avoids the complexity of creating a custom cleanup hook for useMemo. I'll add this modification.

[the remark this answers is not in the document]
Comment (jeangq24) [reply src/react/CozyEventProvider.tsx]: [abe464f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/abe464f622653421397566dc48409119d3f6c2dc)

File: README.md [hunk 2]
```diff
@@ -156,6 +156,261 @@ eventEmitter.destroy();
...
+
+#### `useCozyEvent`
+
+The `useCozyEvent` hook allows components to subscribe to specific events emitted by the `CozyEvent` instance provided by the nearest `CozyEventProvider`.
+
+**Parameters:**
+
+- `eventName` (string): The name of the event to subscribe to.
+- `callback` (function): The function to execute when the event is emitted.
+- `namespace` (optional, string): A namespace to scope the event subscription.
```
[the remark this answers is not in the document]
Comment (jeangq24) [reply README.md]: [a34eb67](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/a34eb67174f384c6b9049c25a58f05a210cca014)

File: src/react/CozyEventProvider.tsx [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * If not provided, a global instance is used by default.
+ * @param {React.ReactNode} children - The child components that will have access to the CozyEvent instance.
+ * @returns {JSX.Element} A React context provider wrapping the children with the CozyEvent instance.
+ * @throws {Error} If the provided instance is not a valid CozyEvent instance.
+ */
+export const globalCozyEventInstance = new CozyEvent(); // Use the global instance if there is no context
+export const CozyEventProvider: FC<CozyEventProviderProps> = ({
+  instance = globalCozyEventInstance,
+  children,
+  id = 'default', // Default ID for the provider
```
Comment (RecursiveVoid) [inline src/react/CozyEventProvider.tsx:23]: please remove the comment here

File: src/react/useCozyEvent.tsx [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  callback: (data: any) => void,
+  options?: UseCozyEventOptions
+) => {
+  
+  const { namespace, id } = options || {};
+  const emitter = id
+  ? getCozyEventInstanceById(id)
+  : useContext(CozyEventContext) || globalCozyEventInstance;
+
+  // Validate the emitter
```
Comment (RecursiveVoid) [inline src/react/useCozyEvent.tsx:48]: please remove the comment here

File: src/react/useCozyEvent.tsx [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+  if (!emitter) {
+    throw new Error(`No CozyEvent instance found for id: ${id}`);
+  }
+
+  // Validate eventName
```
Comment (RecursiveVoid) [inline src/react/useCozyEvent.tsx:53]: please remove the comment here

File: src/react/useCozyEvent.tsx [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  if (typeof eventName !== 'string' || eventName.trim() === '') {
+    throw new Error('Invalid eventName provided to useCozyEvent. It must be a non-empty string.');
+  }
+
+  // Validate callback
```
Comment (RecursiveVoid) [inline src/react/useCozyEvent.tsx:58]: please remove the comment here

File: src/react/useCozyEvent.tsx [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+  if (typeof callback !== 'function') {
+    throw new Error('Invalid callback provided to useCozyEvent. It must be a function.');
+  }
+
+  // Memorize the full event name
```
Comment (RecursiveVoid) [inline src/react/useCozyEvent.tsx:63]: please remove the comment here

File: src/index.ts [hunk 8]
```diff
@@ -1 +1,7 @@
+// Export core CozyEvent functionality
 export { CozyEvent } from './core/CozyEvent';
+// Export React-specific functionality
```
Comment (RecursiveVoid) [inline src/index.ts:3]: please remove the comment here

**Review by RecursiveVoid — CHANGES_REQUESTED** [review]
There are couple of comments that should be removed. other than that, we look like okay i guess.

File: src/react/CozyEventProvider.tsx (hunk 3 above)
Comment (jeangq24) [reply src/react/CozyEventProvider.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/react/CozyEventProvider.tsx [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+}) => {
+  // Validate the instance
```
[the remark this answers is not in the document]
Comment (jeangq24) [reply src/react/CozyEventProvider.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/react/CozyEventProvider.tsx [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+
+   // Register and unregister the instance with useEffect
+   useEffect(() => {
+    registerCozyEventInstance(id, instance);
+    return () => {
+      unregisterCozyEventInstance(id);
+    };
+  }, [id, instance]);
+  
+  // Memorize the context value to avoid unnecessary re-renders
```
[the remark this answers is not in the document]
Comment (jeangq24) [reply src/react/CozyEventProvider.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/react/CozyEventProvider.tsx [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+  instance = globalCozyEventInstance,
+  children,
+  id = 'default', // Default ID for the provider
+}) => {
+  // Validate the instance
+  if (!(instance instanceof CozyEvent)) {
+    throw new Error('Invalid CozyEvent instance provided to CozyEventProvider');
+  }
+
+   // Register and unregister the instance with useEffect
```
[the remark this answers is not in the document]
Comment (jeangq24) [reply src/react/CozyEventProvider.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/react/useCozyEvent.tsx (hunk 4 above)
Comment (jeangq24) [reply src/react/useCozyEvent.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/react/useCozyEvent.tsx (hunk 7 above)
Comment (jeangq24) [reply src/react/useCozyEvent.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/react/useCozyEvent.tsx [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+  const fullEventName = useMemo(
+    () => (namespace ? `${namespace}:${eventName}` : eventName),
+    [namespace, eventName]
+  );
+
+  // Memorize the callback to avoid unnecessary re-registrations
```
[the remark this answers is not in the document]
Comment (jeangq24) [reply src/react/useCozyEvent.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/react/CozyEventProvider.tsx [hunk 13]
```diff
@@ -0,0 +1,46 @@
...
+ * to its child components. It ensures that the provided instance is valid and accessible
+ * throughout the component tree.
+ *
+ * @param {CozyEvent} [instance=globalCozyEventInstance] - An optional instance of CozyEvent. 
+ * If not provided, a global instance is used by default.
+ * @param {React.ReactNode} children - The child components that will have access to the CozyEvent instance.
+ * @returns {JSX.Element} A React context provider wrapping the children with the CozyEvent instance.
+ * @throws {Error} If the provided instance is not a valid CozyEvent instance.
+ */
+export const globalCozyEventInstance = new CozyEvent(); // Use the global instance if there is no context
```
[the remark this answers is not in the document]
Comment (jeangq24) [reply src/react/CozyEventProvider.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/index.ts [hunk 14]
```diff
@@ -1 +1,7 @@
+// Export core CozyEvent functionality
```
[the remark this answers is not in the document]
Comment (jeangq24) [reply src/index.ts]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/index.ts (hunk 8 above)
Comment (jeangq24) [reply src/index.ts]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/react/useCozyEvent.tsx (hunk 5 above)
Comment (jeangq24) [reply src/react/useCozyEvent.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

File: src/react/useCozyEvent.tsx (hunk 6 above)
Comment (jeangq24) [reply src/react/useCozyEvent.tsx]: [322fb5f](https://github.com/RecursiveVoid/CozyEvent/pull/6/commits/322fb5f69f7520a2829949732a0fca35ee93e417)

**Review by RecursiveVoid — APPROVED** [review]
